File: HW1/im_python/Product.py
class Product:
    def __init__(self, section, category, name, price, quantity, qualities=None):
        self.section = section
        self.category = category
        self.name = name
        self.price = price
        self.quantity = quantity
        self.qualities = qualities if qualities else {}
        
    def update_stock(self, add_quantity):
        self.quantity += add_quantity
        print(f"{self.name} stock updated to {self.quantity}")
        
    def display_info(self):
        qualities_str = ", ".join([f"{value}" for key, value in self.qualities.items()])
        return (
            f"{self.name} - ${self.price:.2f} - {self.quantity} - {qualities_str}"
        )
        
    def purchase(self, quantity, discount=0):
        if quantity <= 0:
            print(f"Please enter valid number of items")
            return None
        
        if quantity > self.quantity:
            print(f"Insufficient Stock. Only {self.quantity} avaliable")
            return None
        
        # Discount based on quantity
        if quantity >= 5:
            discount = 0.10  #10% off per item if 5 or more items are purchased
        elif quantity >= 10:
            discount = 0.20  #20% off per item if 10 or more items are purchased
        
        discounted_price = self.price * (1 - discount)
        total_cost = discounted_price * quantity
        undiscounted_cost = self.price * quantity
        
        self.update_stock(-quantity)
        print(f"Purchased {quantity} x {self.name} - Total Price: ${total_cost:.2f}")
        print(f"Total Savings: ${undiscounted_cost - total_cost}")

File: HW1/im_cpp/Product.cpp
#include "Product.h"

Product::Product(string section, string category, string name, double price, int quantity, map<string, string> qualities)
    : section(section), category(category), name(name), price(price), quantity(quantity), qualities(qualities) {}

void Product::update_stock(int add_quantity) {
    quantity += add_quantity;
    cout << name << " stock updated to " << quantity << endl;
}

string Product::display_info() const{
    string qualities_str = "";
    for (const auto& pair : qualities) {
        qualities_str += pair.second + ", ";
    }
    if (!qualities_str.empty()) {
        qualities_str.pop_back();
        qualities_str.pop_back();
    }

    return name + " - $" + to_string(price) + " - " + to_string(quantity) + " - " + qualities_str;
}

void Product::purchase(int quantity_to_buy, double discount) {
    if (quantity_to_buy <= 0) {
        cout << "Please enter a valid number of items" << endl;
        return;
    }

    if (quantity_to_buy > quantity) {
        cout << "Insufficient Stock. Only " << quantity << " available" << endl;
        return;
    }

    if (quantity_to_buy >= 10) {
        discount = 0.20;
    } else if (quantity_to_buy >= 5) {
        discount = 0.10;
    }

    double discounted_price = price * (1 - discount);
    double total_cost = discounted_price * quantity_to_buy;
    double undiscounted_cost = price * quantity_to_buy;

    update_stock(-quantity_to_buy);

    cout << "Purchased " << quantity_to_buy << " x " << name << " - Total Price: $" << total_cost << endl;
    cout << "Total Savings: $" << (undiscounted_cost - total_cost) << endl;
}